device/pcie/pci_device.cpp

```diff
@@ -493,3 +493,53 @@ void PCIDevice::write_tlb_reg(uint32_t byte_addr, std::uint64_t value_lower, std
     }
     tt_driver_atomics::mfence(); // Otherwise subsequent WC loads move earlier than the above UC store to the TLB register.
 }
+
+// Get TLB index (from zero), check if it's in 16MB, 2MB or 1MB TLB range, and dynamically program it.
+dynamic_tlb PCIDevice::set_dynamic_tlb(unsigned int tlb_index, tt_xy_pair start, tt_xy_pair end,
+                            std::uint64_t address, bool multicast, std::unordered_map<chip_id_t, std::unordered_map<tt_xy_pair, tt_xy_pair>>& harvested_coord_translation, std::uint64_t ordering) {
+    auto architecture_implementation = get_architecture_implementation();
+    if (multicast) {
+        std::tie(start, end) = architecture_implementation->multicast_workaround(start, end);
+    }
+
+    log_trace(LogSiliconDriver, "set_dynamic_tlb with arguments: tlb_index = {}, start = ({}, {}), end = ({}, {}), address = 0x{:x}, multicast = {}, ordering = {}",
+         tlb_index, start.x, start.y, end.x, end.y, address, multicast, (int)ordering);
+
+    tt::umd::tlb_configuration tlb_config = architecture_implementation->get_tlb_configuration(tlb_index);
+    std::uint32_t TLB_CFG_REG_SIZE_BYTES = architecture_implementation->get_tlb_cfg_reg_size_bytes();
+    auto translated_start_coords = harvested_coord_translation.at(logical_id).at(start);
+    auto translated_end_coords = harvested_coord_translation.at(logical_id).at(end);
```

Comment:
I don't have anything actionable to suggest here, just some thoughts - 

Passing this in seems clumsy.  I don't understand all of the coordinate stuff, but if this structure is going to live inside the PCIDevice, it doesn't need to be keyed off of logical id anymore (each PCIDevice can have its own).  Also I think the Metal guys would be happy if this were built out of array lookups rather than hash table.

Reply:
I'll fix this when removing logical_id from pci_device. We should at least be passing harvested_coord for this chip only. But also probably not even that (that will be left for even further refactor)